### PR TITLE
Task#26 - Feat: Handling Cart Logic 

### LINKED ISSUE #95: User Story#2: Manage products in shopping cart
> As a user, I want to be able to manage products in my shopping cart, including adding products to the cart and removing products from the cart.

- #96

- [x] Add an "Add to Cart" button with a clear label and an appropriate icon to represent adding a product to the cart.
- [x] Add a "Remove from Cart" button with a clear label and a suitable icon to signify removing a product from the cart.
- [x] Define the appearance of the buttons, including color and layout, for a user-friendly experience.

- #98

  - Implement the logic for adding a product to the cart when the "Add to Cart" button is clicked:
    - [x] Retrieve the selected product's details, such as ID, name, price, and quantity.
    - [x] Check if the product is already in the cart. If so, update the quantity; otherwise, add the product to the cart.
    - [x] Update the cart indicator to reflect the new total number of items.
  
  -  Implement the logic for removing a product from the cart when the "Remove from Cart" button is clicked:
      - [x]  Identify the product to be removed.
      - [x]  Remove the product from the cart's list of items.
      - [x]  Update the cart indicator to show the updated total number of items.

- #102

- [x] When a product is successfully added to the cart, increment the cart icon indicator to reflect the updated total number of items.
- [x] When a product is removed from the cart, decrement the cart icon indicator to show the revised total number of items.

Task#4:  Test Cart Management Functionality

- [x] Verify that clicking the "Add to Cart" button appropriately adds the selected product to the cart and updates the cart indicator.
- [x] Confirm that clicking the "Remove from Cart" button effectively removes the chosen product from the cart and adjusts the cart indicator accordingly.
- [x] Test edge cases, such as adding and removing multiple quantities of the same product, as well as different products consecutively.

### LINKED ISSUE #98: Task#26 - DevTask: Implement Cart Handling Logic
- #89 
  - #95 
## OVERVIEW
- Implement the logic for adding a product to the cart when the "Add to Cart" button is clicked:
  - When the user clicks on the add to cart button, the cart will update that product
  - The total of the cart will be updated with the quantity each time a product is added

- Implement the logic for removing a product from the cart when the "Remove from Cart" button is clicked:
  - When the user clicks on the button with an X on the item in the cart, the product will be removed from the cart
  - The total of the cart will be updated with the quantity each time a product is removed
## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/113186169/789708ff-0a73-437b-a989-a2494b08b2e4

### LINKED ISSUE #102: Task#27 - DevTask: Update Cart Status Indicator
- #89 
  - #95 
## OVERVIEW
- Update message show when add to cart
- [x]  When a product is successfully added to the cart, increment the cart icon indicator to reflect the updated total number of items.

## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/27277174/178568ca-5e8b-4403-8df7-de3905f76830

### PR DESCRIPTION
**Pull request by TheHienDuong**
- Closes #98 
- Closes #102 

## OVERVIEW
- Update button components
- Create button add to cart
- Handling add and remove products in the cart
- Update onAddToCart props to components
- Handling display of products in the cart
- Display total products in the total section of the shopping cart
- Create Message components to show notification when add to cart

## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/113186169/7ba09ffb-6eac-42fc-b212-72c1ca30fe76

### CODE REVIEW

File: practice-two/src/App.jsx [hunk 1]
```diff
@@ -86,6 +89,26 @@ const App = () => {
     });
   };
 
+  /**
+   * Toggle the products in cart
+   *
+   * @param {string} productID - ID of the product to toggle products cart
+   * @returns {Array} - Updated list of products cart IDs
+   */
+  const toggleProductCart = (productID) => {
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: Check naming here. What is the action? I don't understand toggle action

File: practice-two/src/components/CartItem.jsx [hunk 2]
```diff
@@ -68,11 +69,15 @@ const CartItem = ({ product = [] }) => {
     setQuantity(quantity + 1);
   };
 
+  const handleProductCart = () => {
```
Comment (phattruong-agilityio) [inline practice-two/src/components/CartItem.jsx]: check naming

File: practice-two/src/components/ProductDetails.jsx [hunk 3]
```diff
@@ -42,6 +44,10 @@ const ProductDetails = ({
     onToggleProductFavorite(product.id);
   };
 
+  const handleAddToCart = () => {
+    onAddToCart(product.id);
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx:48]: product is [] so why we have `product.id` 
image

File: practice-two/src/components/ProductDetails.jsx [hunk 4]
```diff
@@ -98,12 +104,21 @@ const ProductDetails = ({
...
                 </Typography>
-                <Button
-                  variant="outline"
-                  customVariant="bg-secondary hover:bg-blue-800 focus:bg-blue-800 py-2 px-4 rounded text-white text-xl font-semibold text-sm justify-center"
-                >
-                  Buy It Now
-                </Button>
+                <div className="flex flex-col lg:flex-row gap-5">
+                  <Button
+                    variant="custom"
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx]: custom ?? Check naming here

File: practice-two/src/components/ProductDetails.jsx [hunk 5]
```diff
@@ -98,12 +104,21 @@ const ProductDetails = ({
...
+                    customClass="bg-secondary hover:bg-blue-800 focus:bg-blue-800"
+                  >
+                    Buy It Now
+                  </Button>
+                  <Button
+                    variant="custom"
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ProductDetails.jsx]: The same here

File: practice-two/src/components/ShoppingCart.jsx [hunk 6]
```diff
@@ -18,27 +18,30 @@ import { ICON } from '../constants/data';
  * When the button is clicked, a SliderBar with a list of  products is displayed.
  *
  * @param products - An array of all products
+ * @param cartProductIDs - An array of IDs of products in the cart
+ * @param onAddToCart - Function to handle adding a product to the cart
  * @returns {JSX.Element} The ShoppingCart Component
  */
-const ShoppingCart = ({ products = [] }) => {
+const ShoppingCart = ({ products = [], cartProductIDs = [], onAddToCart = () => {} }) => {
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ShoppingCart.jsx]: check code style

File: practice-two/src/components/ShoppingCart.jsx [hunk 7]
```diff
@@ -53,29 +56,37 @@ const ShoppingCart = ({ products = [] }) => {
...
+              <div className="mt-20 flex w-full flex-col items-center justify-center overflow-hidden">
+                <img src={ICON.cart} alt="cart-icon" className="w-24" />
+                <Typography variant="plain" className="mt-6 text-center text-2xl font-bold">
+                  Your cart is empty.
+                </Typography>
+              </div>
+            ) : (
+              <div className="flex h-full w-full flex-col justify-between overflow-hidden p-1 pb-10">
+                <ul className="flex-grow overflow-auto py-4">
+                  {cartProductIDs.map((productId) => {
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ShoppingCart.jsx:75]: Why do find in map and return null ??? Could you please refactor it

File: practice-two/src/components/common/Button.jsx [hunk 8]
```diff
@@ -9,6 +9,7 @@ const variantClasses = {
   primary: `${BUTTON.DEFAULT} rounded-lg`,
   secondary: `${BUTTON.DEFAULT} rounded-[50px]`,
   outline: 'border-none',
+  custom: `${BUTTON.DEFAULT} py-2 px-4 w-[100%] lg-[50%] rounded text-white text-xl font-semibold justify-center`,
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/Button.jsx]: does this one is difference from primary or secondary ?

Comment (phattruong-agilityio) [inline practice-two/src/components/common/Button.jsx]: does this one is difference from primary or secondary ?

File: practice-two/src/components/common/Button.jsx [hunk 9]
```diff
@@ -40,7 +41,7 @@ const Button = ({
   size = 'medium',
   type = 'button',
   color = 'light',
-  customVariant = '',
+  customClass = '',
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/Button.jsx]: check naming here. It's plural or not

File: practice-two/src/components/common/Popover.jsx [hunk 10]
```diff
@@ -49,7 +49,7 @@ const Popover = ({
             {children}
             {closeButton && (
               <div className="absolute top-1 right-2">
-                <Button customVariant={closeButtonStyle} onClick={onClosePopover}>
+                <Button customClass={closeButtonStyle} onClick={onClosePopover}>
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/Popover.jsx]: customClass={closeButtonStyle} ??
Check naming here

File: practice-two/src/components/common/SliderBar.jsx [hunk 11]
```diff
@@ -41,7 +41,7 @@ const SliderBar = ({
           {children}
           {closeButton && (
             <div className="absolute top-1 right-2">
-              <Button customVariant={closeButtonStyle} onClick={onCloseSliderBar}>
+              <Button customClass={closeButtonStyle} onClick={onCloseSliderBar}>
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/SliderBar.jsx]: customClass={closeButtonStyle} ??
Check naming here

File: practice-two/src/components/layout/FilterCategoryLayout.jsx [hunk 12]
```diff
@@ -82,7 +86,7 @@ const FilterCategoryLayout = ({
           />
           <Button onClick={onResetFilters}>Reset Filters</Button>
         </div>
-        <ShoppingCart products={products} />
+        <ShoppingCart onAddToCart={onAddToCart} products={products} cartProductIDs={cartProductIDs} />
```
Comment (phattruong-agilityio) [inline practice-two/src/components/layout/FilterCategoryLayout.jsx]: check code style

File: practice-two/src/components/layout/FilterCategoryLayout.jsx [hunk 13]
```diff
@@ -123,6 +127,8 @@ FilterCategoryLayout.propTypes = {
     manufacturer: PropTypes.arrayOf(PropTypes.string).isRequired,
     price: PropTypes.arrayOf(PropTypes.string).isRequired,
   }),
+  cartProductIDs: PropTypes.arrayOf(PropTypes.string).isRequired,
```
Comment (phattruong-agilityio) [inline practice-two/src/components/layout/FilterCategoryLayout.jsx:135]: Check this type
image

File: practice-two/src/App.jsx (hunk 1 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Done

File: practice-two/src/components/CartItem.jsx (hunk 2 above)
Comment (TheHienDuong) [reply practice-two/src/components/CartItem.jsx]: Updated

File: practice-two/src/components/ProductDetails.jsx (hunk 3 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx:48]: There seems to be an inconsistency in the code. The product prop is given a default value of an empty array ( product = []) in the function parameter, but in the code it is treated as an object with properties like product.id, which would lead to error. Changed default value for products

File: practice-two/src/components/ProductDetails.jsx (hunk 4 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx]: Yes, Updated into 'purchase'

File: practice-two/src/components/ProductDetails.jsx (hunk 5 above)
Comment (TheHienDuong) [reply practice-two/src/components/ProductDetails.jsx]: Yes, Updated into 'purchase'

File: practice-two/src/components/ShoppingCart.jsx (hunk 7 above)
Comment (TheHienDuong) [reply practice-two/src/components/ShoppingCart.jsx:75]: Yes, Removed

File: practice-two/src/components/common/Button.jsx (hunk 8 above)
Comment (TheHienDuong) [reply practice-two/src/components/common/Button.jsx]: Updated

Comment (TheHienDuong) [reply practice-two/src/components/common/Button.jsx]: Updated

File: practice-two/src/components/common/Button.jsx (hunk 9 above)
Comment (TheHienDuong) [reply practice-two/src/components/common/Button.jsx]: Yes, Updated into 'customClasses'

File: practice-two/src/components/common/SliderBar.jsx (hunk 11 above)
Comment (TheHienDuong) [reply practice-two/src/components/common/SliderBar.jsx]: Updated

File: practice-two/src/components/layout/FilterCategoryLayout.jsx (hunk 12 above)
Comment (TheHienDuong) [reply practice-two/src/components/layout/FilterCategoryLayout.jsx]: Updated

File: practice-two/src/components/layout/FilterCategoryLayout.jsx (hunk 13 above)
Comment (TheHienDuong) [reply practice-two/src/components/layout/FilterCategoryLayout.jsx:135]: Done

File: practice-two/src/components/ShoppingCart.jsx (hunk 6 above)
Comment (TheHienDuong) [reply practice-two/src/components/ShoppingCart.jsx]: Checked, but formatting pulls them up 1 row

File: practice-two/src/components/ButtonFavorite/ButtonFavorite.jsx [hunk 14]
```diff
@@ -38,7 +38,7 @@ const FavoriteButton = ({ onClick = () => {}, isFavorite = false }) => {
 
   return (
     <Button
-      customVariant="relative w-[70px] h-[46px] -top-3 "
+      className="relative w-[70px] h-[46px] -top-3 "
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ButtonFavorite/ButtonFavorite.jsx]: ??

Comment (phattruong-agilityio) [reply practice-two/src/components/ButtonFavorite/ButtonFavorite.jsx]: customClasses? or className? Please use one and correct

File: practice-two/src/components/ShoppingCart.jsx (hunk 7 above)
Comment (phattruong-agilityio) [reply practice-two/src/components/ShoppingCart.jsx:75]: ??? still not correct

File: practice-two/src/components/common/Button.jsx [hunk 15]
```diff
@@ -2,13 +2,14 @@ import React from 'react';
 import PropTypes from 'prop-types';
 
 const BUTTON = {
-  DEFAULT: 'hover:border-secondary border-2 border-light',
+  LIGHT: 'hover:border-secondary border-2 border-light',
+  COLOR: 'rounded text-white text-xl font-semibold justify-center',
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/Button.jsx]: check naming here

Comment (phattruong-agilityio) [reply practice-two/src/components/common/Button.jsx]: What is COLOR ?

File: practice-two/src/App.jsx (hunk 1 above)
Comment (phattruong-agilityio) [reply practice-two/src/App.jsx]: I don't see any update

File: practice-two/src/components/CartItem.jsx (hunk 2 above)
Comment (phattruong-agilityio) [reply practice-two/src/components/CartItem.jsx]: handleProductCart?? Handle must be do something. Check naming here

File: practice-two/src/components/ShoppingCart.jsx (hunk 7 above)
Comment (phattruong-agilityio) [reply practice-two/src/components/ShoppingCart.jsx:75]: check it plz

File: practice-two/src/components/common/Popover.jsx [hunk 16]
```diff
@@ -23,7 +23,7 @@ const Popover = ({
   onClosePopover = () => {},
   closeButton = true,
   closeButtonContent = 'Close',
-  closeButtonStyle = '',
+  customCloseButton = '',
```
Comment (phattruong-agilityio) [inline practice-two/src/components/common/Popover.jsx]: i really don't understand why we need this Props

**Comment by phattruong-agilityio** [comment]
When user click "Add to card" I don't see any respond from application. Could you please update
Here is example

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/27277174/178568ca-5e8b-4403-8df7-de3905f76830
